dashboard/src/components/AppView/AppControls/RollbackButton/RollbackButton.test.tsx

```diff
@@ -0,0 +1,146 @@
+import { shallow, ShallowWrapper } from "enzyme";
+import context from "jest-plugin-context";
+import * as React from "react";
+import * as ReactModal from "react-modal";
+import { IAppRepository, IChartVersion, IRelease } from "shared/types";
+import RollbackButton from ".";
+import SelectRepoForm from "../../../../components/SelectRepoForm";
+import RollbackDialog from "./RollbackDialog";
+
+const defaultProps = {
+  app: {} as IRelease,
+  rollbackApp: jest.fn(),
+  getChartVersion: jest.fn(),
+  loading: false,
+  repos: [],
+  repo: {} as IAppRepository,
+  kubeappsNamespace: "kubeapps",
+  fetchRepositories: jest.fn(),
+  checkChart: jest.fn(),
+};
+
+it("stores the chart name and version in the state", () => {
+  const wrapper = shallow(<RollbackButton {...defaultProps} />);
+  wrapper.setProps({
+    app: {
+      chart: {
+        metadata: {
+          name: "foo",
+          version: "1.0.0",
+        },
+      },
+    } as IRelease,
+  });
+
+  expect(wrapper.state()).toMatchObject({ chartName: "foo", chartVersion: "1.0.0" });
+});
+
+function openModal(wrapper: ShallowWrapper) {
+  ReactModal.setAppElement(document.createElement("div"));
+  wrapper.setState({ modalIsOpen: true });
+  wrapper.update();
+}
+
+context("when opening the Modal", () => {
+  context("when there is no info about the chart", () => {
+    it("should render the SelectRepoForm", () => {
+      const wrapper = shallow(<RollbackButton {...defaultProps} chart={undefined} />);
+      openModal(wrapper);
+
+      expect(wrapper.find(SelectRepoForm)).toExist();
+      expect(wrapper.find(RollbackDialog)).not.toExist();
+    });
+
+    it("should fetch the available repositories", () => {
+      const fetchRepositories = jest.fn();
+      const wrapper = shallow(
+        <RollbackButton
+          {...defaultProps}
+          chart={undefined}
+          fetchRepositories={fetchRepositories}
+        />,
+      );
+      const button = wrapper.find(".button");
+      expect(button).toExist();
+      button.simulate("click");
+      expect(fetchRepositories).toBeCalled();
+    });
+
+    it("if the app has updateInfo, fetch the chart", () => {
+      const getChartVersion = jest.fn();
+      const app = {
+        updateInfo: {
+          repository: {
+            name: "foo",
+          },
+        },
+      } as IRelease;
+      const wrapper = shallow(
+        <RollbackButton
+          {...defaultProps}
+          chart={undefined}
+          getChartVersion={getChartVersion}
+          app={app}
+        />,
+      );
+      wrapper.setState({ chartName: "bar", chartVersion: "1.0.0" });
```

Comment:
It's better to test how the component reacts to user interactions and lifecycle methods than forcing states manually. By forcing the state you're skipping the natural component behaviour and it may skip bugs from user interactions or other conditions.

Reply:
In this case the `setState` is not needed but yes, I try to reproduce users behavior when possible. In some cases (not this one) I cannot do so because simulations like "onConfirm" don't work as expected.